api/envoy/config/wasm/v2/wasm.proto

```diff
@@ -0,0 +1,68 @@
+syntax = "proto3";
+
+package envoy.config.wasm.v2;
+
+option java_outer_classname = "WasmProto";
+option java_multiple_files = true;
+option java_package = "io.envoyproxy.envoy.config.wasm.v2";
+
+import "envoy/api/v2/core/base.proto";
+
+import "validate/validate.proto";
+
+// [#protodoc-title: Wasm service]
+
+// Configuration for a Wasm VM.
+// [#next-free-field: 6]
+message VmConfig {
+  // An ID which will be used along with a hash of the wasm code (or null_vm_id) to determine which
+  // VM will be used for the plugin. All plugins which use the same vm_id and code will use the same
+  // VM. May be left blank.
+  string vm_id = 1;
+
+  // The Wasm runtime type (see source/extensions/common/wasm/well_known_names.h).
+  string runtime = 2;
+
+  // The Wasm code that Envoy will execute.
+  api.v2.core.AsyncDataSource code = 3;
+
+  // The Wasm configuration string used on initialization of a new VM (proxy_onStart).
```

Comment:
What does this even do?

Reply:
We separated out VM configuration from Plugin configuration.  The former is only provided once when the VM starts up and should contain general configuration (e.g. metrics sinks) where the latter are provided when a plugin (e.g. a filter in a filter chain) is created or when the configuration is updated dynamically.  The plugin configuration should contain plugin specific information.   Note that the VM might outlive the filter chain (because it is relatively expensive and if the code doesn't change we can continue using the VM).   We can push plugin configuration updates independently from the lifetime of the VM or of the filter/plugin.

Reply:
The lifetime here is pretty confusing, I think we need to work through some examples. Imagine we have a listener with a bunch of WASM filters and VMs. They have state, which gets updated over time, e.g. via FCDS or the future filter specific config xDS. Then we drain a listener and launch a new one for the same port. What happens if the VMs overlap? How is configuration created/destroyed at the VM level? What about in-flight plugins?

Reply:
I sent out the document describing the lifetime.  Basically, the VM is reused if 1) the code is the same (SHA256 of wasm bytecode) and 2) the vm_id hasn't changed.   The former because it has to be recreated to get the new code and the latter to provide a way to force reloading of the VM (i.e. increment a version vm_id). 

Two VMs can overlap in the same way that two listeners overlap.  When a listener is draining, no new requests/streams are created, so whatever the current configuration is will continue to be used.  For a new VM, dynamic configuration arrival can update the configuration in the VM which can be snapped by the plugin author at their discretion.  A best practices document is in order.

Do you have a suggestion for the location and format of such documents?

We can have them in envoy/api/wasm/... or envoy /docs/root/wasm/... for example.

Reply:
We need to have arch overview docs for WASM that are user visible. If there are dev only docs they can be markdown. 
For end user arch overview docs I would put then here: https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/other_features/scripting (maybe consider moving scripting to a top level arch overview topic and then having Lua and WASM sections)

For dev docs put them here: https://github.com/envoyproxy/envoy/tree/master/source/docs

Reply:
(Also I didn't see any document sent out)

Reply:
The architecture doc was sent out.  We have https://github.com/envoyproxy/envoy/issues/9462 for the end user documentation and https://github.com/envoyproxy/envoy/issues/9465 for developer information.